### PR TITLE
[Master] Fix13526 allow JSX attributes to be union type

### LINKED ISSUE #13526: JSX element attributes type […] may not be a union type
**TypeScript Version:**  2.1.1

**Code**
As seen here: http://stackoverflow.com/questions/41682619/jsx-element-attributes-type-may-not-be-a-union-type/41683341#41683341.

```ts
// address.tsx
...
interface Address {
  street: string;
  country: string;
}

interface CanadianAddress extends Address {
  postalCode: string;
}

interface AmericanAddress extends Address {
  zipCode: string;
}

type Properties = AmericanAddress | CanadianAddress;

function isAmerican(address: Properties) address is AmericanAddress {
 return 'zipCode' in address;
}

export class Address extends React.Component<Properties, void> {
  public render() {
    let isAmerican = isAmerican(this.props.address);
    ...
  }
}

// map.tsx
...
let rootDiv = document.createElement('root')l
ReactDOM.render(<Address postalCode='T1B 0L3' />, rootDiv);
```

**Expected behavior:**
Should be able to specify a union type for a React Component's properties generic variable.

**Actual behavior:**
```
error TS2600: JSX element attributes type '({ children?: ReactNode; } & AmericanAddress) | ({ children?: ReactNode; } & CanadianAddress' may not be a union type.
```

Clearly this is an anticipated use case, is there a specific reason why this can't be done, and if so, is there any sane way to get around this? Or do I simply need to use optional attributes in a single interface with custom error logic inside the component to disallow both `zipCode` and `postalCode` being specified at the same time?

### PR DESCRIPTION
**Pull request by yuit**
Fix #13526

### CODE REVIEW

File: tests/baselines/reference/tsxAttributeResolution16.js [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+type Properties = CanadianAddress | AmericanAddress;
+
+export class AddressComp extends React.Component<Properties, void> {
+  public render() {
+    return null;
+  }
+}
+
+let a = <AddressComp postalCode='T1B 0L3' street="vancouver" country="CA" />
```
Comment (sandersn) [inline tests/baselines/reference/tsxAttributeResolution16.js]: this postal code is in Medicine Hat not Vancouver

Comment (yuit) [inline tests/baselines/reference/tsxAttributeResolution16.js]: 🤣

**Comment by lndigo** [comment]
Any idea what I'm doing wrong?
[error](http://imgur.com/a/eVVIk)

**Comment by mhegazy** [comment]
What version of TS are you using ?

**Comment by lndigo** [comment]
2.3.0-dev.20170222
I left it in the screenshot to verify that I have it loaded for the current workspace.

**Comment by yuit** [comment]
@lndigo  the merge is just got in today. The fix will be in tomorrow Typescript@next

**Comment by lndigo** [comment]
I don't believe it is working with higher order components. I'm getting error TS2322, type xxx is not assignable to type yyy.

Version: 2.3.0-dev.20170223

```ts
interface Address {
  street: string;
  country: string;
}

interface CanadianAddress extends Address {
    postalCode: string;
}

interface AmericanAddress extends Address {
    zipCode: string;
}

type Properties = CanadianAddress | AmericanAddress;

interface AddressHocProps {
    properties:Properties;
}

export class AddressComp extends React.Component<Properties, void> {
  public render() {
    return null;
  }
}

export class AddressHoc extends React.Component<AddressHocProps,void>{
    render(){
        //error ts2322
        return <AddressComp {...this.props.properties} />
    }
}
```

**Pull request merged**